app/assets/javascripts/3dbio_viewer/src/data/repositories/BionotesPdbInfoRepository.ts

```diff
@@ -3,182 +3,260 @@ import { FutureData } from "../../domain/entities/FutureData";
 import { Maybe } from "../../utils/ts-utils";
+import { getStorageCache, setStorageCache } from "../storage-cache";
 import i18n from "../../domain/utils/i18n";
-import { getSessionCache, setSessionCache } from "../session-cache";
 
 export class BionotesPdbInfoRepository implements PdbInfoRepository {
-    get(pdbId: PdbId, onProcessDelay: (reason: string) => void): FutureData<PdbInfo> {
-        const proteinMappingUrl = `${routes.bionotes}/api/mappings/PDB/Uniprot/${pdbId}`;
-        const fallbackProteinMappingUrl = `${routes.ebi}/pdbe/api/mappings/uniprot/${pdbId}`;
-        const polymerCoverage = `${routes.ebi}/pdbe/api/pdb/entry/polymer_coverage/${pdbId}/`;
-        const emdbMapping = `${emdbsFromPdbUrl}/${pdbId}`;
+    get(args: GetPdbInfoArgs): FutureData<PdbInfo> {
+        const { pdbId } = args;
 
-        const bionotesProteinMapping$ = getFromUrl<Maybe<BioUniprotFromPdbMapping>>(
-            proteinMappingUrl
-        ).flatMapError(
-            (_err): FutureData<Maybe<BioUniprotFromPdbMapping>> => Future.success(undefined)
-        );
+        const data$ = {
+            uniprotMapping: this.getBionotesProteinMapping(pdbId),
+            fallbackProteinMapping: this.getEbiProteinMapping(pdbId),
+            molecules: this.getPolymerCoverage(pdbId),
+            emdbMapping: getFromUrl<PdbEmdbMapping>(`${emdbsFromPdbUrl}/${pdbId}`),
+        };
 
-        const ebiProteinMapping$ = getFromUrl<Maybe<EbiUniprotFromPdbMapping>>(
-            fallbackProteinMappingUrl
-        ).flatMapError(
-            (_err): FutureData<Maybe<EbiUniprotFromPdbMapping>> => Future.success(undefined)
-        );
+        return Future.joinObj(data$).flatMap(data => this.getPdbInfo(data, args));
+    }
 
-        const polymerCoverage$ = getFromUrl<ChainsFromPolymer>(polymerCoverage)
-            .flatMap(
-                (polymerCoverage): FutureData<PolymerMolecules> => {
-                    const molecules = polymerCoverage[pdbId.toLowerCase()]?.molecules;
+    private getPdbInfo(data: Data, args: GetPdbInfoArgs): FutureData<PdbInfo> {
+        const { pdbId, onProcessDelay } = args;
+        const { uniprotMapping, fallbackProteinMapping, emdbMapping, molecules } = data;
 
-                    if (!molecules)
-                        return buildError("noData", {
-                            message: "Polymer coverage not found for this PDB.",
-                        });
-                    else return Future.success(molecules);
-                }
-            )
-            .flatMapError(err => buildError("noData", err));
+        const hasProteinRes = uniprotMapping || fallbackProteinMapping;
+        if (!hasProteinRes) console.debug(`Uniprot mapping not found for ${pdbId}`);
 
-        const emdbMapping$ = getFromUrl<PdbEmdbMapping>(emdbMapping);
+        const chains = molecules
+            .flatMap(({ chains }) => chains)
+            .map(chain => ({
```

Comment:
We are building a `ChainIdMapping`, right? in that case, let's be explicit about it: `(chain): ChainIdMapping`. This not only helps TS giving better error message, it also link the structure to the type, so we get automatic renaming of props, code navigation, and so on.